### PR TITLE
Implement conversion to complex scalar

### LINKED ISSUE #782: complex(cupy.array(1j)) not working
Conversion to complex scalar is not implemented for `cupy.ndarray`. It fails while calling into `__float__`.

### PR DESCRIPTION
**Pull request by kohr-h**
Closes #782

### CODE REVIEW

**Comment by okuta** [comment]
Cloud you add test?

**Comment by kohr-h** [comment]
Sure, test added.

**Comment by okuta** [comment]
Thank you!

**Comment by okuta** [comment]
jnekins, test this please

**Comment by okuta** [comment]
I got error in Python3.
Cloud you fix it?
```
13:49:07 =================================== FAILURES ===================================
13:49:07 _________________ TestScalarConversion.test_scalar_conversion __________________
13:49:07 
13:49:07 self = <cupy_tests.core_tests.test_ndarray_complex_ops.TestScalarConversion testMethod=test_scalar_conversion>
13:49:07 dtype = <class 'numpy.complex64'>
13:49:07 
13:49:07     @testing.for_all_dtypes()
13:49:07     def test_scalar_conversion(self, dtype):
13:49:07         scalar = 1 + 1j if numpy.dtype(dtype).kind == 'c' else 1
13:49:07         x_1d = cupy.array([scalar]).astype(dtype)
13:49:07 >       self.assertEqual(complex(x_1d), x_1d[0])
13:49:07 
13:49:07 dtype      = <class 'numpy.complex64'>
13:49:07 scalar     = (1+1j)
13:49:07 self       = <cupy_tests.core_tests.test_ndarray_complex_ops.TestScalarConversion testMethod=test_scalar_conversion>
13:49:07 x_1d       = array([ 1.+1.j], dtype=complex64)
13:49:07 
13:49:07 tests/cupy_tests/core_tests/test_ndarray_complex_ops.py:88: 
13:49:07 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:49:07 
13:49:07 >   return complex(self.get())
13:49:07 E   TypeError: can't convert complex to float
```

**Comment by kohr-h** [comment]
Hm, no idea why I didn't get this error. I pushed a fix that works locally.

**Comment by okuta** [comment]
jenkins, test this please.

File: tests/cupy_tests/core_tests/test_ndarray_complex_ops.py [hunk 1]
```diff
@@ -72,3 +74,18 @@ def test_imag_setter_raise(self, xp, dtype):
...
         return x
+
+
+@testing.gpu
+class TestScalarConversion(unittest.TestCase):
+
+    _multiprocess_can_split_ = True
+
+    @testing.for_all_dtypes()
+    def test_scalar_conversion(self, dtype):
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_ndarray_complex_ops.py]: Cloud you add `@testing.with_requires('numpy>=1.12.0')`?
That error caused by old NumPy (<1.12.0).

Comment (kohr-h) [reply tests/cupy_tests/core_tests/test_ndarray_complex_ops.py]: Ah right, Numpy has no `__complex__` before 1.12. I'll fix it.

Comment (kohr-h) [reply tests/cupy_tests/core_tests/test_ndarray_complex_ops.py]: Done

Comment (okuta) [reply tests/cupy_tests/core_tests/test_ndarray_complex_ops.py]: Thanks

**Comment by okuta** [comment]
jnekins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by kohr-h** [comment]
Does it work now?

**Comment by okuta** [comment]
Sorry, late reply.
The test was passed. LGTM!

**Pull request merged**